{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IMPORTS\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram, fcluster, cut_tree\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import os\n",
    "import sys\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from lifelines import KaplanMeierFitter\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "\n",
    "\n",
    "sys.path.append(os.getcwd().strip('notebooks') + 'src/')\n",
    "import WGCNA_functions as wgcnax\n",
    "\n",
    "\n",
    "import importlib\n",
    "importlib.reload(wgcnax)\n",
    "\n",
    "\n",
    "import dill\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### LOADING REAL UNPUBLISHED DATA    -     NO PUSHING FOR THE RESULTS\n",
    "\n",
    "# Move out of the notebook folder to access datasets\n",
    "working_dir = os.getcwd()\n",
    "working_dir = working_dir.strip('notebooks')\n",
    "data_dir = working_dir + 'data/PROTECTED_DATA/BGI_Expression_Data/'\n",
    "\n",
    "\n",
    "## Load the dataset\n",
    "# Transcriptomics Data \n",
    "transcriptomics_TumorOnly_dir = data_dir + 'CRC.SW.mRNA.symbol.TPM_TumorOnly.csv'\n",
    "transcriptomics_dataset = pd.read_csv(transcriptomics_TumorOnly_dir, index_col=0)\n",
    "\n",
    "# Sample info and Clinical Traits Data\n",
    "CT_for_TumorSamples_dir = data_dir + 'ClinicalTraits_for_TumorSamples.csv'\n",
    "trait_dataset = pd.read_csv(CT_for_TumorSamples_dir, index_col=0)\n",
    "\n",
    "SurvivalData_for_TumorSamples_dir = data_dir + 'Survival_Data_for_TumorSamples.csv' # Using only tumor samples\n",
    "survival_dataset = pd.read_csv(SurvivalData_for_TumorSamples_dir, index_col=0)\n",
    "\n",
    "\n",
    "# Figures Saving output dir\n",
    "figures_dir = working_dir + 'results/HC_simple/'\n",
    "\n",
    "# Check if the directory exists, and if not, create it\n",
    "if not os.path.exists(figures_dir):\n",
    "    os.makedirs(figures_dir)\n",
    "    print(f\"Creating directory to save results and figures...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### BASIC DENDOGRAM USING CORRELATOIN AS THE DISTANCE \n",
    "dpi_general = 100\n",
    "module_member_threshold = 0.25   ## Set the threshold value to consider members of a module.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## We check for zero variance genes & Nan\n",
    "variances = transcriptomics_dataset.var()  # Compute variance for each column\n",
    "zero_var_columns = variances[variances == 0].index \n",
    "transcriptomics_dataset_filtered = transcriptomics_dataset.drop(columns=zero_var_columns)\n",
    "print(\"Columns with zero variance being removed:\", zero_var_columns.tolist())\n",
    "\n",
    "\n",
    "## Compute the metrics for the dendogram based on clustering genes\n",
    "distances = pdist(transcriptomics_dataset_filtered.T, metric='correlation')\n",
    "linkage_matrix =  linkage(distances, method=\"average\")\n",
    "\n",
    "\n",
    "## Plot the dendrogram\n",
    "plt.figure(figsize=(15, 7))\n",
    "dendrogram(linkage_matrix, truncate_mode=None, color_threshold=module_member_threshold, \n",
    "           labels=transcriptomics_dataset_filtered.T.index, leaf_rotation=90)\n",
    "title_general_dendogram = 'General Hierarchical Clustering Dendrogram'\n",
    "plt.title(title_general_dendogram)\n",
    "plt.xlabel('Genes')\n",
    "plt.ylabel('Distance (as a measure of correlation)')\n",
    "plt.tight_layout()\n",
    "plt.savefig(figures_dir + title_general_dendogram, dpi=100)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Flat clusters\n",
    "\n",
    "min_num_members = 5\n",
    "max_num_members = 150\n",
    "\n",
    "percentile = 0.9\n",
    "dpi_modules = 100   \n",
    "\n",
    "\n",
    "## Identify clusters (modules) in the dendogram\n",
    "cluster_assignation = fcluster(linkage_matrix, t=percentile, criterion='distance')\n",
    "unique_clusters = np.unique(cluster_assignation)\n",
    "corrected_cluster_assignation = np.zeros(cluster_assignation.shape, dtype=int)\n",
    "\n",
    "\n",
    "# Assign new cluster IDs, filtering based on the size criteria\n",
    "new_cluster_id = 1\n",
    "for cluster_id in unique_clusters:\n",
    "    members = np.where(cluster_assignation == cluster_id)[0]\n",
    "    if min_num_members <= len(members) <= max_num_members:\n",
    "        corrected_cluster_assignation[members] = new_cluster_id\n",
    "        new_cluster_id += 1\n",
    "\n",
    "\n",
    "print(f\"{len(np.unique(corrected_cluster_assignation))} modules where identified, fulfilling the restrictions\")\n",
    "\n",
    "module_assignment = pd.DataFrame({\n",
    "    'Gene Name': transcriptomics_dataset_filtered.columns,\n",
    "    'Module': corrected_cluster_assignation\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot Cluster representation\n",
    "\n",
    "fig = plt.figure(figsize=(16, 12))\n",
    "\n",
    "# Histogram of the number of genes in each module, excluding not assigned (Module 0)\n",
    "ax1 = fig.add_subplot(2, 1, 1)\n",
    "genes_in_modules = module_assignment[module_assignment['Module'] != 0]['Module'].value_counts().sort_index()\n",
    "ax1.bar(genes_in_modules.index, genes_in_modules.values)\n",
    "ax1.set_title('Number of Genes in Each Module')\n",
    "ax1.set_xlabel('Module')\n",
    "ax1.set_ylabel('Number of Genes')\n",
    "\n",
    "# Setup for pie charts\n",
    "ax2 = fig.add_subplot(2, 2, 3)\n",
    "ax3 = fig.add_subplot(2, 2, 4)\n",
    "\n",
    "# Pie chart of genes distribution across modules\n",
    "module_sizes = module_assignment['Module'].value_counts()\n",
    "ax2.pie(module_sizes, labels=module_sizes.index, autopct='%1.1f%%', startangle=140)\n",
    "ax2.set_title('Distribution of Genes Across Modules')\n",
    "\n",
    "# Percentage of Genes Assigned to Clusters vs Not Assigned\n",
    "total_genes = len(module_assignment)\n",
    "genes_not_assigned_count = module_assignment[module_assignment['Module'] == 0].shape[0]\n",
    "genes_assigned_count = total_genes - genes_not_assigned_count\n",
    "sizes = [genes_assigned_count, genes_not_assigned_count]\n",
    "labels = ['Assigned to Modules', 'Not Assigned (Module 0)']\n",
    "ax3.pie(sizes, labels=labels, autopct='%1.1f%%', startangle=140)\n",
    "ax3.set_title('Percentage of Genes Assigned to Clusters vs Not Assigned')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(figures_dir + 'ModuleDistribution', dpi=100)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Build a full expression profile for all the genes, keeping the module assignation\n",
    "\n",
    "transcriptomics_data_transposed = transcriptomics_dataset_filtered.T\n",
    "transcriptomics_data_transposed.reset_index(inplace=True)\n",
    "transcriptomics_data_transposed.rename(columns={'index': 'Gene Name'}, inplace=True)\n",
    "\n",
    "\n",
    "expression_profiles = pd.merge(module_assignment, transcriptomics_data_transposed, on='Gene Name', how='left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(f\"Calculating EigenGenes...\")\n",
    "\n",
    "eigengenes = []\n",
    "\n",
    "## Iterate through each module to calculate its eigen gene\n",
    "for module in expression_profiles['Module'].unique():\n",
    "\n",
    "    # Skip module 0 as it represents unassigned genes\n",
    "    if module == 0:\n",
    "        continue\n",
    "    else:\n",
    "        # Extract the Expression Profile for all genes in this module\n",
    "        module_expression_profile = expression_profiles[expression_profiles['Module'] == module].iloc[:, 2:]\n",
    "\n",
    "        # Perform PCA on the expression data of the current module\n",
    "        pca = PCA(n_components=1)\n",
    "        pca.fit(module_expression_profile)\n",
    "        \n",
    "        # The first principal component is the eigengene\n",
    "        eigengene = pca.components_[0]\n",
    "\n",
    "        # Create a DataFrame for the eigengene with the correct sample labels and the module id\n",
    "        eigengene_df = pd.DataFrame(eigengene.reshape(1, -1), columns=expression_profiles.columns[2:])\n",
    "        eigengene_df.insert(0, 'Module', module)\n",
    "        eigengenes.append(eigengene_df)\n",
    "\n",
    "eigengenes = pd.concat(eigengenes, ignore_index=True)\n",
    "\n",
    "\n",
    "## Plot the Expression Profile for the Eigengenes across pacients\n",
    "print(f\"Plotting and Saving the Eigengene Expression Profile Across Samples...\")\n",
    "title_figure = 'Eigengene Expression Profile Across Samples'\n",
    "\n",
    "sample_labels = eigengenes.columns[1:]\n",
    "\n",
    "plt.figure(figsize=(15, 10))\n",
    "for index, row in eigengenes.iterrows():\n",
    "    # Convert eigengene array stored as list back to numpy array for plotting\n",
    "    eigengene_values = np.array(row[1:].values)\n",
    "    \n",
    "    # Plotting the eigengene values\n",
    "    plt.plot(sample_labels, eigengene_values, label=f'Module {row[\"Module\"]}')\n",
    "\n",
    "plt.title(title_figure, fontsize=20)\n",
    "plt.xlabel('Samples (pacients)', fontsize=10)\n",
    "plt.ylabel('Eigengene Expression Level', fontsize=10)\n",
    "plt.xticks(rotation=90)\n",
    "plt.xticks([])\n",
    "#plt.legend()\n",
    "plt.tight_layout()\n",
    "plt.savefig(figures_dir + title_figure, dpi=100)\n",
    "plt.show()\n",
    "print(f\"Done\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Step 8: Module-Trait Relationship\n",
    "# Run the Analysis, encoding the variables as categorical, and calculating correlation and p-value\n",
    "\n",
    "# THIS USES THE NEW ENCODING AND CORRELATION FUNCTIONS\n",
    "trait_columns = list(trait_dataset.columns[1:] )\n",
    "correlations, p_values = wgcnax.eigen_trait_correlations_DC(eigengenes, trait_dataset, trait_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wgcnax.correlation_pvalue_heatmap(correlations, p_values, figures_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find rows where all cells have abs(value) < 0.4 and drope them\n",
    "threshold_of_interest = 0.4\n",
    "\n",
    "# Filter out non relevant modules\n",
    "not_significant_modules = (correlations.abs() < threshold_of_interest).all(axis=1)\n",
    "modules_to_drop = correlations.index[not_significant_modules]\n",
    "print(f\"The following Modules are droped: {list(modules_to_drop)}\")\n",
    "correlations_filtered = correlations.drop(index=modules_to_drop)\n",
    "p_values_filtered = p_values.drop(index=modules_to_drop)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot the Heatmap for the module-trait relationship\n",
    "\n",
    "#### Filtered Heatmap, displaying only reelevant clusters\n",
    "\n",
    "wgcnax.correlation_pvalue_heatmap(correlations_filtered, p_values_filtered, figures_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "significant_modules = []\n",
    "for trait in correlations.columns:\n",
    "    significant_clusters = np.abs(correlations[trait]) > threshold_of_interest\n",
    "    significant_clusters = significant_clusters[significant_clusters].index.tolist()\n",
    "    if significant_clusters:\n",
    "        for module in significant_clusters:\n",
    "            significant_modules.append(module)\n",
    "            print(f'Trait: {trait}  correlates with Module: {module}, showing a correlation of {correlations.at[module, trait]:.3f}')\n",
    "\n",
    "print('\\n')\n",
    "for module in set(significant_modules):\n",
    "    module_expression_profile = expression_profiles[expression_profiles['Module'] == module]\n",
    "    module_Genes = module_expression_profile['Gene Name'].tolist()\n",
    "    print(f'The module with id:{module} clusters {len(module_Genes)} genes.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Step 9: Survival plot\n",
    "\n",
    "# Get the modules that show high correlations\n",
    "wgcnax.survival_probability(correlations, threshold_of_interest, expression_profiles, survival_dataset, figures_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot a Network Representation of the Relevant Clusters\n",
    "\n",
    "# Color mapping for different intervals of correlation strengths\n",
    "interval_colors = {\n",
    "    '0.9 - 1.0': 'red',\n",
    "    '0.8 - 0.9': 'orange',\n",
    "    '0.7 - 0.8': 'yellow',\n",
    "    '0.6 - 0.7': 'blue',\n",
    "    \n",
    "    '0.0 - 0.5': 'green',\n",
    "    }\n",
    "\n",
    "# Function to determine color based on weight\n",
    "def get_edge_color(weight, interval_colors):\n",
    "    for interval, color in interval_colors.items():\n",
    "        lower, upper = map(float, interval.split(' - '))\n",
    "        if lower <= abs(weight) <= upper:\n",
    "            return color\n",
    "    return 'grey'\n",
    "\n",
    "\n",
    "num_modules = len(set(significant_modules))\n",
    "grid_size = int(np.ceil(np.sqrt(num_modules)))\n",
    "plt.figure(figsize=(grid_size * 5, grid_size * 5))\n",
    "for i, cluster in enumerate(set(significant_modules), start = 1):\n",
    "    # Build Expression Profile for the Module\n",
    "    module_profile = expression_profiles[expression_profiles['Module'] == cluster].copy()\n",
    "    module_profile.set_index('Gene Name', inplace=True)\n",
    "    module_profile.drop('Module', axis=1, inplace=True)\n",
    "\n",
    "    # Add Eigengene\n",
    "    module_eigengene = eigengenes[eigengenes['Module'] == cluster].copy()\n",
    "    module_eigengene.set_index('Module', inplace=True)\n",
    "    module_eigengene.rename(index={cluster: 'EigenGene'}, inplace=True)\n",
    "    module_profile = pd.concat([module_profile, module_eigengene])\n",
    "\n",
    "    # Convert to numpy and calculate Spearman correlation -> Membership function in a way\n",
    "    module_profile_matrix = module_profile.to_numpy()\n",
    "    corr, _ = spearmanr(module_profile_matrix, axis=1)\n",
    "    corr_matrix = pd.DataFrame(corr, index=module_profile.index, columns=module_profile.index)\n",
    "    \n",
    "    # Network creation and node addition\n",
    "    G = nx.Graph()\n",
    "    for gene in module_profile.index:\n",
    "        G.add_node(gene)\n",
    "    \n",
    "    # Edge addition based on correlation\n",
    "    for gene1 in module_profile.index:\n",
    "        for gene2 in module_profile.index:\n",
    "            if gene1 != gene2:\n",
    "                weight = corr_matrix.loc[gene1, gene2]\n",
    "                color = get_edge_color(weight, interval_colors)\n",
    "                G.add_edge(gene1, gene2, weight=weight, color=color)\n",
    "    \n",
    "    # Plotting the network in a subplot\n",
    "    plt.subplot(grid_size, grid_size, i)\n",
    "    pos = nx.spring_layout(G, k=0.5, iterations=20)\n",
    "    edges = G.edges(data=True)\n",
    "    nx.draw(G, pos, with_labels=True, node_color=['red' if name=='EigenGene' else 'skyblue' for name in G.nodes], \n",
    "            node_size=500, \n",
    "            edge_color=[data['color'] for _, _, data in edges], linewidths=1, font_size=10,\n",
    "            width=[(data['weight'])*2 for _, _, data in edges])\n",
    "\n",
    "    plt.title(f'Module {cluster}')\n",
    "    plt.axis('off')\n",
    "\n",
    "\n",
    "\n",
    "# Add a legend for the whole figure\n",
    "legend_labels = {label: plt.Line2D([0], [0], color=color, lw=4) for label, color in interval_colors.items()}\n",
    "plt.figlegend(legend_labels.values(), legend_labels.keys(), loc='upper right')\n",
    "plt.tight_layout()\n",
    "plt.suptitle('Network Representation of all Relevant Modules', fontsize=20)\n",
    "plt.tight_layout(rect=[0, 0.03, 1, 0.95])\n",
    "\n",
    "plt.savefig(figures_dir + title_figure, dpi=200)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save session \n",
    "session_file = working_dir + 'data/Sessions/Hierarchical_Clustering_session.pkl'\n",
    "dill.dump_session(session_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Move out of the notebook folder to access datasets\n",
    "working_dir = os.getcwd()\n",
    "working_dir = working_dir.strip('notebooks')\n",
    "\n",
    "# Load session \n",
    "session_file = '/media/bernat/Work/Hierarchical_Clustering_session.pkl'\n",
    "dill.load_session(session_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "WGCNA_approach",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
